### ISSUE TITLE
L.FeatureGroup.SubGroup is not a constructor

### ISSUE THREAD

**Issue by nkalinov** [ghybs/Leaflet.FeatureGroup.SubGroup#8]
Hi @ghybs,
Thanks for your plugins.

My "setup" (included in this order):
- **leaflet@1.0.0-rc.2** 
- **leaflet.markercluster@1.0.0-rc.1.0**
- **leaflet.featuregroup.subgroup@1.0.0**:

I have the following error:
`L.FeatureGroup.SubGroup is not a constructor`

**Comment by ghybs** [comment]
Hi,

Thank you for your feedback! :-)

Please can you provide more details on your context and issue?
Share some code, ideally reproduce your issue on an online editing service (Plunker, JSFiddle, JSBin, etc.)

I am not able to reproduce it: https://jsfiddle.net/z92L4y17/1/

**Comment by nkalinov** [comment]
Sorry @ghybs,
Actually I need the **Leaflet.MarkerCluster.LayerSupport** plugin (which works just fine), not this one.... 
I probably had some build/bundling issues with this one, not sure why.
For the info:
Tried to use it with Ionic2 / Typescript project.
Imported the packages in app.ts:

```
import 'leaflet';
import 'leaflet.markercluster';
import './leaflet.featuregroup.subgroup';
```

Thanks again!

**Comment by jhenligne** [comment]
Hi @ghybs,

I also want to give your plugin a try and face the same problem.
Config:
leaflet-1.0.0-rc.1.js
leaflet.markercluster-1.0.0-rc.1.0.js

It may not be related to your code because I am using browserify, (> 'browserify test.js -o bundle.js').

Copying and pasting your https://jsfiddle.net/z92L4y17/4/ code result in that same error: L.FeatureGroup.SubGroup is not a constructor.

html and js files provided for tests.

[test.zip](https://github.com/ghybs/Leaflet.FeatureGroup.SubGroup/files/482826/test.zip)

**Comment by ghybs** [comment]
Hi @jhenligne,

Thank you for your report.

Please can you try changing in the UMD section at the top of the dist file:

``` javascript
define(["leaflet"], function (L) {
  return (root.L.FeatureGroup.SubGroup = factory(L));
});
```

by removing the reference to `root`: `return (L.FeatureGroup.SubGroup = factory(L));`

``` javascript
define(["leaflet"], function (L) {
  return (L.FeatureGroup.SubGroup = factory(L));
});
```

Please can you come back and tell if it made any difference?

That is the main difference in the module definition between SubGroup and MCG.LayerSupport, which nkalinov sounds to be using without that issue.

**Comment by jhenligne** [comment]
Hi @ghybs,

Meanwhile I did like @nkalinov and it works well with MCG.LayerSupport.

Going back to Leaflet.FeatureGroup.SubGroup and removing root reference I still get same kind of error with FG.SubGroup.

It is thrown here:

``` javascript
// Supply with a factory for consistency with Leaflet.
L.featureGroup.subGroup = function (parentGroup, options) {
    return new FG.SubGroup(parentGroup, options);
};
```

**Comment by jhenligne** [comment]
Sorry, there is an error in my previous post.

I did this modification in my own code and noticed a difference with your Jsfiddle.

Writing `subgroup = new L.FeatureGroup.SubGroup(mcg);` throws "L.FeatureGroup.SubGroup is not a function" and all the expression is underlined in red in Chrome console,
writing `subgroup = L.FeatureGroup.SubGroup(mcg);` throws same error but only ".SubGroup(mcg);" is underlined in red.

Writing `subgroup = L.featureGroup.subGroup(mcg);` (notice lower caps: **f**eatureGroup.**s**ubGroup) raise exception describe in my previous post. This syntax is the one given in your example.

**Comment by jhenligne** [comment]
OK, I found the problem.

Your last lines are:

``` javascript
// Supply with a factory for consistency with Leaflet.
L.featureGroup.subGroup = function (parentGroup, options) {
    return new FG.SubGroup(parentGroup, options);
};
```

It should be:

``` javascript
// Supply with a factory for consistency with Leaflet.
L.featureGroup.subGroup = function (parentGroup, options) {
    return new SubGroup(parentGroup, options);
};
```

As SubGroup extends FG, there is no need to make a reference to it.

**Comment by ghybs** [comment]
Awesome!

Please was the removal of the reference to `root` in `return (root.L.FeatureGroup.SubGroup = factory(L));` necessary as well?
(actually it looks like this line would be for AMD / Require.js, rather than CommonJS / Browserify)

**Comment by jhenligne** [comment]
No, it didn't help. That's why I continued digging.

**Comment by ghybs** [comment]
Excellent, thank you for your troubleshooting!

Please could you create a PR to fix that issue?

**Comment by jhenligne** [comment]
Yes, will do later.

**Comment by ghybs** [comment]
Looks like the fix is like that one: https://github.com/ghybs/Leaflet.TileLayer.Fallback/pull/1

I wonder the same thing as in that PR, i.e. in CommonJS world, would it be expected (or forbidden on the contrary) to have a side effect?

Typically here to assign `L.FeatureGroup.SubGroup` so that later you can call either the factory `L.featureGroup.subGroup()` or the constructor `new L.FeatureGroup.SubGroup()`?

Well, thinking further about it, the factory `L.featureGroup.subGroup` is already a side effect, so I guess it would not hurt to do it as well for `L.FeatureGroup.SubGroup`… :-) Probably in the `build/umd.template.js` file I guess, like what is done for the browser environment.

**Comment by ghybs** [comment]
For the sake of completeness, it looks like I should have stuck to Leaflet's recommended [plugin wrapper template](https://github.com/Leaflet/Leaflet/blob/master/PLUGIN-GUIDE.md#module-loaders) (which is actually more a variation of the [jQuery plugin wrapper template](https://github.com/umdjs/umd/blob/master/templates/jqueryPlugin.js)) rather than the official UMD template for [returnExportsGlobal](https://github.com/umdjs/umd/blob/master/templates/returnExportsGlobal.js), even though it is listed as the reference by the Leaflet Plugin Guide… :-S

**Comment by jhenligne** [comment]
Interesting remarks.

It's quite late here so I won't do it before tomorrow.
If you don't want to wait please do the fix yourself.

**Comment by ghybs** [comment]
Please could you try with the version in that Plunker?
https://plnkr.co/edit/mm9ZmiFpzEtZmkre27SR?p=preview
(go to the "leaflet.featuregroup.subgroup-src.js" file on the top left "FILES" pane)

I simplified the UMD wrapper so that there is now only side effects, but I am not sure if it works right with CommonJS / Browserify?

**Comment by jhenligne** [comment]
Hi @ghybs,
It's working fine this way with my code.

Just a question, why are you prefixing your class with it's parent name?
For example `MarkerClusterGroup` also extends `L.FeatureGroup` but is not name `L.FeatureGroup.MarkerClusterGroup`.
I suppose you want to make obvious the fact it is a subclass but it may confuse some users and give them trouble. The way it is called shows already it may be used as a substitute of `L.FeatureGroup`.

**Comment by ghybs** [comment]
Hi @jhenligne,

Thank you a lot for your confirmation!
I will update the releases, and eventually my other plugins as well...

As for the lengthy name, I do not really like it either, but this is to comply with Leaflet [plugin guide](https://github.com/Leaflet/Leaflet/blob/master/PLUGIN-GUIDE.md#plugin-api). Of course there is no enforcement for compliance, but at least it shows my respect for Leaflet authors... :-)

As for Leaflet.markercluster, I like the short name as well. My understanding is that the name was chosen when Leaflet was still young. Besides, it no longer really behaves like a Feature Group (e.g. it does not emit layer add events, it does not keep track of child groups, etc.)

**Comment by jhenligne** [comment]
Good to know we have the same feeling. Thank you for your explanation.

I am happy this little bug won't refrain people to use your plugin anymore because I find it useful. It should be included in core Leaflet.

**Comment by ghybs** [comment]
Just to let you know that I released version `v1.0.1` with that fix, and I took the time to learn how to publish to npm :-)
https://www.npmjs.com/package/leaflet.featuregroup.subgroup
`npm i leaflet.featuregroup.subgroup --save`

**Comment by jhenligne** [comment]
Good work.
Thank you.
